### PR TITLE
Handle intersection types when looking up base types for visibility

### LINKED ISSUE #25264: Bug: tsc crash TypeError: Cannot read property 'flags' of undefined
**TypeScript Version:**  2.9.2 and `typescript@next` at the time of writing.

**Search Terms:** TypeError: Cannot read property 'flags' of undefined

**Code**

```ts
class Abstract {
  protected constructor() {
  }
}

class Concrete extends Abstract {
}

type Constructor<T = {}> = new (...args: any[]) => T;

function Mixin<TBase extends Constructor>(Base: TBase) {
  return class extends Base {
  };
}

class Empty {
}

class CrashTrigger extends Mixin(Empty) {
  public trigger() {
    new Concrete();
  }
}
```
Example repo with code to reproduce: https://github.com/purificant/typescript-crash-example

**Expected behavior:**
tsc emits an error along the lines of "TS2674: Constructor of class 'Abstract' is protected and only accessible within the class declaration."

**Actual behavior:**
tsc crashes with the following stack trace:
```
typescript-crash-example/node_modules/typescript/lib/tsc.js:63860
                throw e;
                ^

TypeError: Cannot read property 'flags' of undefined
    at getBaseTypes (typescript-crash-example/node_modules/typescript/lib/tsc.js:25683:38)
    at isConstructorAccessible (typescript-crash-example/node_modules/typescript/lib/tsc.js:36070:37)
    at resolveNewExpression (typescript-crash-example/node_modules/typescript/lib/tsc.js:36024:22)
    at resolveSignature (typescript-crash-example/node_modules/typescript/lib/tsc.js:36180:28)
    at getResolvedSignature (typescript-crash-example/node_modules/typescript/lib/tsc.js:36209:26)
    at checkCallExpression (typescript-crash-example/node_modules/typescript/lib/tsc.js:36282:29)
    at checkExpressionWorker (typescript-crash-example/node_modules/typescript/lib/tsc.js:37687:28)
    at checkExpression (typescript-crash-example/node_modules/typescript/lib/tsc.js:37629:42)
    at checkExpressionStatement (typescript-crash-example/node_modules/typescript/lib/tsc.js:39545:13)
    at checkSourceElement (typescript-crash-example/node_modules/typescript/lib/tsc.js:41180:28)
```

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #25264

`getBaseTypes` returns a `BaseType[]` (which is a `(InterfaceType | IntersectionType)[]`), but inside `isConstructorAccessible`, we were casting away and discarding the intersection type (aka mixin) case.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -19073,11 +19073,22 @@ namespace ts {
                     let baseTypes = getBaseTypes(containingType as InterfaceType);
                     while (baseTypes.length) {
                         const baseType = baseTypes[0];
-                        if (modifiers & ModifierFlags.Protected &&
-                            baseType.symbol === declaration.parent.symbol) {
-                            return true;
+                        if (baseType.flags & TypeFlags.Intersection) {
+                            for (const intersectionMember of (baseType as IntersectionType).types) {
+                                if (getObjectFlags(intersectionMember) & (ObjectFlags.Class | ObjectFlags.Interface)) {
+                                    baseTypes = getBaseTypes(intersectionMember as InterfaceType);
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why stop at the first one?

Comment (mhegazy) [reply src/compiler/checker.ts]: it can be an intersection of two types with constructors, right?

Comment (weswigham) [reply src/compiler/checker.ts]: Intersection types technically allow it, but we're using an intersection to approximate a mixin, where that won't be the case. It'll be one interface type (the original base) with one (or more) anonymous object type(s) with the mixin'd members. And we require mixin contructors have `...any[]` constructor signatures if one is present, specifically so we know they don't muck with the expected constructors implied by the base class.

Comment (mhegazy) [reply src/compiler/checker.ts]: but i can have a const whose type is a mixin type.. we filter out the ones we do not like using isMixinConstructorType

Comment (weswigham) [reply src/compiler/checker.ts]: I've refactored the code a bit and explicitly handled mixins.

**Pull request merged**